design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,291 @@
+
+- _Custom Struct Support_ - UDFs can only operate on standard Java types (`Map`, `List`, `String`, 
+etc...) and do not accept structured types. This dramatically limits what KSQL can handle, as
+much of the data flowing through kafka is structured Avro/JSON.
+- _Generics Type Support_ - Today, there is no way to implement a single UDF that supports data 
+structures with implicit types. For example, implementing a generic `list_union` operator would
+require different method signatures for each supported type. This quickly becomes unmanageable
+when there are multiple arguments.
+- _Variable Argument Support_ - Varargs are required to create UDFs that accept an arbitrary
+number of inputs. Creating a simple `concat` UDF that could sum multiple columns would require
+creating method signatures for as many input combinations as desired.
+- _Complex Aggregation Models_ - Today, KSQL only supports tracking a single variable for 
+aggregation and it must mirror the output type. To implement an `avg` UDAF, it is necessary to store 
+both a `count` and a `running_total`.
+
+## Scope
+
+All changes will be made for the new annotation-based UDF interfaces and will not be backported to
+the legacy implementation. The bullet points mentioned above are all in scope, but the following 
+will not be included in this KLIP:
+
+- UDTFs - multiple output operations are more complicated and will require a KLIP of their own
+- Generics/Varargs for UDAFs - the support for Generics/Varargs will only extend to UDFs.
+
+## Value
+
+Basic UDF operations that are supported in SQL languages will now be unblocked for KSQL. Some of
+these are outlined below:
+
+|      Improvement      |                                 Value                                  |
+|:---------------------:|------------------------------------------------------------------------|
+| Generics support      | Unblock `ARRAY_LENGTH`, `ARRAY_SLICE`, `ARRAY_TO_STRING`, etc...       |
+| Varargs support       | `CONCAT`, `MAX`, `MIN` etc...                                          |
+| Complex Aggregation   | `AVG`, `PERCENTILE`, etc...                                            |
+| Structured UDFs       | Arbitrary data handling without intermediary transforms                |
+
+## Public APIS
+
+Some public APIs will be changed, though all will be changed in a backwards compatible fashion:
+
+|      Improvement      |                               API change                               |
+|:---------------------:|------------------------------------------------------------------------|
+| Generics support      | UDF interface will accept inferred generics (e.g. `List<T>`)           |
+| Varargs support       | UDF interface will accept `...` in the method declarations             |
+| Complex Aggregation   | UDAFs may include `VR export(A agg)` to convert aggregated to output   |
+| Structured UDFs       | UDF/UDAF interfaces will accept the Kafka Connect `Struct` type        |
+
```

Comment:
can we consider adding a bullet for making it possible to access the record `context` from inside a UDF/UDAF ? Here i mean `context` in the Kafka Streams sense (see [](https://kafka.apache.org/21/documentation/streams/developer-guide/processor-api.html#accessing-processor-context)). I guess we might consider that unnecessarily powerful and 'unsafe' in some way, in which case the core motivation here is to access the various metadata fields of the current record (offset, partition, timestamp, ...) so perhaps we can consider copying those out into some value object and giving access to that instead of to the actual ProcessorContext.

Reply:
Stateful operations are a serious beast! Unless this is really one of the more requested features (or would dramatically change the way I design the other four) I'd prefer to leave it out of scope for this KLIP. Lots of small KLIPs make me a happy engineer 😄

Reply:
OOps, my link got screwed up there: https://kafka.apache.org/21/documentation/streams/developer-guide/processor-api.html#accessing-processor-context
Understood! Am asking here in case you feel the need to change the public interface in order to allow for it in some way that would be a PITA or require some compatibility-breaking change later. 
My motivating example for this is a simple `Earliest` or `Latest` UDAF, which returns the specified field from the input record with the lowest or highest incoming timestamp for this window, respectively.

Reply:
Makes sense - I will do some prototyping and thinking about APIs to see if there's any good way without changing public APIs for this. If public APIs need to change, then I agree that we should at least design around that here.

Reply:
@agavra where did we land up on this thread?

Reply:
We would need to change two things:

1. Introduce an API to access it (and likely just a limited subset of information available in it)
2. Change our `SelectValueMapper` to implement `ValueTransformer` instead of `ValueMapper`. Based on my initial exploration, this should be completely backwards compatible; if it is not or adds some performance overhead, we can add some scaffolding code around it to dynamically choose one or the other based on the presence of the API 

Both of these can be easily done in backwards compatible fashion and can probably be done as just a PR. FYI @blueedgenick